## README

These scripts were written for a [blog
post](https://ricomnl.com/blog/bottom-up-bioinformatics-pipeline/)
demonstrating how a bioinformatics pipeline can be implemented. I go through
these scripts in an attempt to learn more about Python scripting.

### 01 digest protein

This is the first script of the pipeline and as the name suggests, it digests a
protein. The protein sequence is stored in a single entry FASTA file and an
enzyme regex should be provided to tell the script how it should be digested.

The full usage can be obtained by typing `./01_digest_protein.py -h`. This
usage is generated by using the
[argparse](https://docs.python.org/3/library/argparse.html) module and I have
written [some
notes](https://davetang.github.io/bioinformatics_tips/get_option.html#Python)
on using this module.

```bash
usage: 01_digest_protein.py [-h] [--enzyme_regex ENZYME_REGEX]
                            [--missed_cleavages MISSED_CLEAVAGES]
                            [--min_length MIN_LENGTH]
                            [--max_length MAX_LENGTH]
                            input_file output_file

positional arguments:
  input_file            A .fasta file containing a protein.
  output_file           A .txt output file to write the peptides to.

optional arguments:
  -h, --help            show this help message and exit
  --enzyme_regex ENZYME_REGEX
                        A regex for the enzyme to use for the digest. E.g.
                        [KR] for trypsin.
  --missed_cleavages MISSED_CLEAVAGES
                        Number of missed cleavages for the digest.
  --min_length MIN_LENGTH
                        Minimum length for a peptide to be considered valid.
  --max_length MAX_LENGTH
                        Maximum length for a peptide to be considered valid.
```

The input to the script is a single entry FASTA file. Below is the amino acid
sequence for [MYC](https://www.genecards.org/cgi-bin/carddisp.pl?gene=MYC)
stored in `MYC.fasta`.

```
>sp|P01106|MYC_HUMAN Myc proto-oncogene protein OS=Homo sapiens OX=9606 GN=MYC PE=1 SV=1
MPLNVSFTNRNYDLDYDSVQPYFYCDEEENFYQQQQQSELQPPAPSEDIWKKFELLPTPP
LSPSRRSGLCSPSYVAVTPFSLRGDNDGGGGSFSTADQLEMVTELLGGDMVNQSFICDPD
DETFIKNIIIQDCMWSGFSAAAKLVSEKLASYQAARKDSGSPNPARGHSVCSTSSLYLQD
LSAAASECIDPSVVFPYPLNDSSSPKSCASQDSSAFSPSSDSLLSSTESSPQGSPEPLVL
HEETPPTTSSDSEEEQEDEEEIDVVSVEKRQAPGKRSESGSPSAGGHSKPPHSPLVLKRC
HVSTHQHNYAAPPSTRKDYPAAKRVKLDSVRVLRQISNNRKCTSPRSSDTEENVKRRTHN
VLERQRRNELKRSFFALRDQIPELENNEKAPKVVILKKATAYILSVQAEEQKLISEEDLL
RKRREQLKHKLEQLRNSCA
```

The script is organised into four functions:

1. `load_fasta`
2. `save_peptides`
3. `digest_protein`
4. `main`

#### Load FASTA

The FASTA file opened using `with open(filename, "r") as fasta_file:` and is
considered [good
practice](https://docs.python.org/3/tutorial/inputoutput.html#reading-and-writing-files)
when dealing with file objects. The entire file is read into `lines` as a long
string and then
[splitlines](https://docs.python.org/3/library/stdtypes.html#str.splitlines) is
used to split the string at line boundaries into a list. The first list entry
is assumed to be FASTA header and every other line afterwards contains the
amino acid sequence and is joined.

#### Save peptides

The list of peptides is written to an output file using `with open(filename,
"w") as output_file:` and by iterating through the list of peptides. The
`write` method is used with
[format](https://docs.python.org/3/library/stdtypes.html#str.format) to save
each peptide into `output_file`. Note that newlines need to be specified with
`write` unlike `print`.

#### Digest protein

The protein sequence is split according to an input regular expression
(abbreviated as regex). The default regex is `[KR]` which defines a _character
class_, which are sets of characters that can match at a single location. This
character class is composed of two amino acids: lysine (K) and arginine (R).
The code `enzyme_regex.finditer(protein_sequence)` finds all amino acids that
are either K or R. The first few matches, start and end index (0 indexed), for
`MYC.fasta` are shown below.

```
9 10
50 51
51 52
64 65
65 66
82 83
125 126
```

The `9 10` indicates that at position 9 until 10, there is a match. We can
confirm this by examining the start of `MYC.fasta`.

```
012345678 9 0123456789
MPLNVSFTN R NYDLDYDSVQ
```

A `sites` list is created to store the matching end index. The `sites` list
starts with a zero index because a single cut downstream of the start of the
sequence can generate a short peptide. Similarily, the `sites` list ends with
the length of the protein sequence because a cut upstream of the end can
generate a peptide.

The function accepts an argument called `missed_cleavages` (default = 0) and
this is used to generate peptide sequences by "skipping" (because a cleavage
was missed) cuts. To illustrate, a missed cleavage of 1 will generate three
peptides for sequence `NNNNNRYYYYYR`:

1. NNNNNR
2. NNNNNRYYYYYR (additional peptide generated due to missed cleavage)
3. YYYYYR

The first five predicted peptides for `MYC.fasta` with a missed cleavage of 0
(default), minimum length of 4 (default), and maximum length of 75 (default)
are shown below.

```
MPLNVSFTNR
NYDLDYDSVQPYFYCDEEENFYQQQQQSELQPPAPSEDIWK
FELLPTPPLSPSR
SGLCSPSYVAVTPFSLR
GDNDGGGGSFSTADQLEMVTELLGGDMVNQSFICDPDDETFIK
```

We can manually check the results by examining the FASTA sequence. The `|`
indicates the match end index, which is used to define the start of the
sequence; thus the start of the peptide sequence does not include the matching
amino acid. The end of the peptide sequence is defined by the next matching end
index.

```
0123456789 | 01234567890123456789012345678901234567890 | 123456789
MPLNVSFTNR | NYDLDYDSVQPYFYCDEEENFYQQQQQSELQPPAPSEDIWK | KFELLPTPP
```

Recall that string slicing does not include the end index. But since the end of
the string slicing index used to generate the peptides is the matching end
index, the matching amino is included as the last peptide.

```python
s = '0123456789'
s[1:5] # returns '1234'
```

#### Main

A main function is used as an [entry
  point](https://en.wikipedia.org/wiki/Entry_point#C_and_C++) and calls the
other three functions.

